src/modules/sih/sih.cpp

```diff
@@ -0,0 +1,687 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file sih.cpp
+ * Simulator in Hardware
+ *
+ * @author Romain Chiappinelli		<romain.chiap@gmail.com>
+ *
+ * Coriolis g Corporation - January 2019
+ */
+
+#include "sih.hpp"
+
+#include <px4_getopt.h>
+#include <px4_log.h>
+
+#include <drivers/drv_pwm_output.h>			// to get PWM flags
+#include <uORB/topics/vehicle_status.h> 	// to get the HIL status
+
+#include <unistd.h> 			//
+#include <string.h>    			//
+#include <fcntl.h> 				//
+#include <termios.h> 			//
+
+using namespace math;
+
+int Sih::print_usage(const char *reason)
+{
+	if (reason) {
+		PX4_WARN("%s\n", reason);
+	}
+
+	PRINT_MODULE_DESCRIPTION(
+		R"DESCR_STR(
+### Simulator in Hardware
+This module provide a simulator for quadrotors running fully 
+inside the hardware autopilot.
+
+This simulator subscribes to "actuator_outputs" which are the actuator pwm
+signals given by the mixer.
+
+This simulator publishes the sensors signals corrupted with realistic noise
+in order to incorporate the state estimator in the loop.
+
+### Implementation
+The simulator implements the equations of motion using matrix algebra. 
+Quaternion representation is used for the attitude.
+Forward Euler is used for integration.
+Most of the variables are declared global in the .hpp file to avoid stack overflow.
+
+
+)DESCR_STR");
+
+	PRINT_MODULE_USAGE_NAME("sih", "sih");
```

Comment:
Second param should be "simulation" as per https://github.com/PX4/Firmware/pull/11400#pullrequestreview-214969363 (which also shows where you'll need to update the category in module docs tools)